include/pybind11/detail/type_caster_base.h

```diff
@@ -827,6 +827,33 @@ using movable_cast_op_type
 template <typename T, typename SFINAE = void>
 struct is_copy_constructible : std::is_copy_constructible<T> {};
 
+// True if Container has a dependent type mapped_type that is equivalent
+// to Container itself
+// Actual implementation in the SFINAE specializations below
+template <typename Container, typename MappedType = Container>
+struct sfinae_is_container_with_self_referential_mapped_type : std::false_type {};
+
+// Tie-breaking between the mapped_type and the value_type specializations is trivial:
+// The specializations are only valid if both conditions are fulfilled:
+// 1) The mapped_type (respectively value_type) exists
+// 2) And it is equivalent to Container
+// So, in each case, only one specialization will activate.
+template <typename Container>
+struct sfinae_is_container_with_self_referential_mapped_type<Container,
+                                                             typename Container::mapped_type>
+    : std::true_type {};
+
+template <typename Container>
+struct sfinae_is_container_with_self_referential_mapped_type<Container,
+                                                             typename Container::value_type>
```

Comment:
Mainly just disclosing that I'm not sure about something: could there ever be a non-non-sensical container that has both `mapped_type` and `value_type`? And what would happen then?

Maybe that's a non-sense thought? And in the worst case we could fix when we actually get into trouble.

But also thinking: do we really need this helper type? Totally untested:

```
struct is_recursive_container<Container, enable_if_t<std::is_same<Container, Container::value_type>::value> : std::true_type;

struct is_recursive_container<Container, enable_if_t<std::is_same<Container, Container::mapped_type>::value> : std::true_type;
```

or

```
struct is_recursive_container<Container, enable_if_t<!std::is_same<Container, Container::value_type>::value && std::is_same<Container, Container::mapped_type>::value> : std::true_type;
```

in case that could ever happen?